service/matching/db.go

```diff
@@ -181,18 +209,57 @@ func (db *taskQueueDB) UpdateState(
 	})
 	if err == nil {
 		db.ackLevel = ackLevel
+		db.emitApproximateBacklogCount()
 	}
 	return err
 }
 
+// updateApproximateBacklogCount updates the in-memory DB state with the given delta value
+func (db *taskQueueDB) updateApproximateBacklogCount(
+	delta int64,
+) {
+	db.Lock()
+	defer db.Unlock()
+
+	// Preventing under-counting
+	if db.approximateBacklogCount.Load()+delta < 0 {
+		db.logger.Info("ApproximateBacklogCounter could have under-counted.",
+			tag.WorkerBuildId(db.queue.BuildId()), tag.WorkerBuildId(db.queue.Partition().NamespaceId().String()))
+		db.approximateBacklogCount.Store(0)
+	} else {
+		db.approximateBacklogCount.Add(delta)
+	}
+}
+
+func (db *taskQueueDB) getApproximateBacklogCount() int64 {
+	return db.approximateBacklogCount.Load()
+}
+
 // CreateTasks creates a batch of given tasks for this task queue
 func (db *taskQueueDB) CreateTasks(
 	ctx context.Context,
-	tasks []*persistencespb.AllocatedTaskInfo,
+	taskIDs []int64,
+	reqs []*writeTaskRequest,
 ) (*persistence.CreateTasksResponse, error) {
 	db.Lock()
 	defer db.Unlock()
-	return db.store.CreateTasks(
+
+	if len(reqs) == 0 {
+		return &persistence.CreateTasksResponse{}, nil
+	}
+
+	maxReadLevel := int64(0)
+	var tasks []*persistencespb.AllocatedTaskInfo
+	for i, req := range reqs {
+		tasks = append(tasks, &persistencespb.AllocatedTaskInfo{
+			TaskId: taskIDs[i],
+			Data:   req.taskInfo,
+		})
+		maxReadLevel = taskIDs[i]
+	}
+	db.approximateBacklogCount.Add(int64(len(tasks)))
```

Comment:
why do we add before and then subtract on some error types? why not add after (if not those error types)?

Reply:
It's because `CreateTasks` should also have the updated task queue metadata right (at-least for non-SQL databases) since it updates the metadata on creation of tasks there?

Reply:
ah right, makes sense